Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/remove_self_intersections.h

```diff
@@ -0,0 +1,1865 @@
+                              VPM& vpm)
+{
+  CGAL::Emptyset_iterator out;
+  replace_faces_with_patch(faces, patch, pmesh, vpm, out);
+}
+
+// -*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-
+
+template <typename Point, typename FaceRange, typename PolygonMesh, typename VertexPointMap>
+void back_up_face_range_as_point_patch(std::vector<std::vector<Point> >& point_patch,
+                                       const FaceRange& face_range,
+                                       const PolygonMesh& tmesh,
+                                       const VertexPointMap vpm)
+{
+  typedef typename boost::graph_traits<PolygonMesh>::halfedge_descriptor    halfedge_descriptor;
+  typedef typename boost::graph_traits<PolygonMesh>::face_descriptor        face_descriptor;
+
+  point_patch.reserve(face_range.size());
+
+  for(const face_descriptor f : face_range)
+  {
+    std::vector<Point> face_points;
+    for(const halfedge_descriptor h : CGAL::halfedges_around_face(halfedge(f, tmesh), tmesh))
+      face_points.push_back(get(vpm, target(h, tmesh)));
+
+    point_patch.push_back(face_points);
+  }
+}
+
+// -*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-
+
+template <typename FaceRange, typename EdgeConstrainMap,
+          typename TriangleMesh, typename VertexPointMap, typename GeomTraits>
+void constrain_sharp_and_border_edges(const FaceRange& faces,
+                                      TriangleMesh& tmesh,
+                                      EdgeConstrainMap& eif,
+                                      const bool constrain_sharp_edges,
+                                      const double dihedral_angle,
+                                      const double /*weak_DA*/,
+                                      VertexPointMap vpm,
+                                      const GeomTraits& gt)
+{
+  typedef typename boost::graph_traits<TriangleMesh>::halfedge_descriptor   halfedge_descriptor;
+  typedef typename boost::graph_traits<TriangleMesh>::edge_descriptor       edge_descriptor;
+  typedef typename boost::graph_traits<TriangleMesh>::face_descriptor       face_descriptor;
+
+  typedef typename GeomTraits::FT                                           FT;
+  typedef typename GeomTraits::Vector_3                                     Vector;
+
+  std::map<edge_descriptor, bool> is_border_of_selection;
+  for(face_descriptor f : faces)
+  {
+    // @fixme what about nm vertices
+    for(halfedge_descriptor h : CGAL::halfedges_around_face(halfedge(f, tmesh), tmesh))
+    {
+      // Default initialization is guaranteed to be `false`. Thus, meet it once will switch
+      // the value to `true` and meeting it twice will switch back to `false`.
+      const edge_descriptor e = edge(h, tmesh);
+      is_border_of_selection[e] = !(is_border_of_selection[e]);
+    }
+  }
+
+#if 0 // Until detect_features++ is integrated
```

Comment:
Update?

Reply:
It is not yet integrated (https://github.com/CGAL/cgal/pull/4084), and I don't like the current detect_features++ enough to push for its integration.